dftimewolf/lib/processors/gcp_cloud_resource_tree.py

```diff
@@ -0,0 +1,1199 @@
+    """Initialize the Cloud Resource Tree Processor.
+
+    Args:
+      state: recipe state.
+      name: The module's runtime name.
+      critical: True if the module is critical, which causes
+          the entire recipe to fail if the module encounters an error.
+    """
+    super(GCPCloudResourceTree, self).__init__(state,
+                                               name=name,
+                                               critical=critical)
+
+    self._project_id = ''
+    self._resource_name = ''
+    self._resource_type = ''
+    self._mode: OperatingMode = OperatingMode.ONLINE
+    self._start_date = ''
+    self._end_date = ''
+    self._period_covered_by_retrieved_logs: Dict[str, datetime] = {}
+    self._resources_dict: Dict[str, Resource] = {}
+
+  # pylint: disable=arguments-differ
+  def SetUp(self, project_id: str, resource_name: str, resource_type: str,
+            mode: str) -> None:
+    """Set up the resource we want to build the tree for.
+
+    Args:
+      project_id: Project id where the resources are located.
+      resource_name: Resource name.
+      resource_type: Resource type (currently supported types: gce_instance,
+        gce_disk, gce_image, gce_machine_image, gce_instance_template,
+        gce_snapshot)
+      mode: operational mode: online or offline
+    """
+    self._project_id = project_id
+    self._resource_name = resource_name
+    self._resource_type = resource_type
+    if 'offline' in mode:
+      self._mode = OperatingMode.OFFLINE
+    elif 'online' in mode:
+      self._mode = OperatingMode.ONLINE
+
+  def Process(self) -> None:
+    """Create the GCP Cloud Resource Tree."""
+    if self._mode == OperatingMode.OFFLINE:
+      self.logger.info('Starting module in offline mode.')
+      # Get the file containers created by the previous module in the recipe
+      file_containers = self.state.GetContainers(containers.File)
+
+      # Loop over the file containers and parse the content of each to fill the
+      # resources dictionary
+      for file_container in file_containers:
+        self.logger.info(
+            f'Loading file from file system container: {file_container.path}')
+        self._ParseLogMessagesFromFileContainer(file_container)
+
+    elif self._mode == OperatingMode.ONLINE:
+      self.logger.info('Starting module in online mode.')
+      self._GetListOfResources(self._project_id)
+      self._GetResourcesMetaDataFromLogs(self._project_id)
+
+    else:
+      self.PublishMessage(
+          'Invalid operating mode. Supported modes are "online" or "offline"')
+
+    self._BuildResourcesParentRelationships()
+
+    matched_resource = self._FindResource(self._resource_name,
+                                          self._resource_type, None,
+                                          self._project_id)
+
+    if not matched_resource:
+      self.logger.error('Resource not found')
+      return
+
+    resource = matched_resource
```

Comment:
Use resource from the beginning, matched_resource variable is redundent

Reply:
Done